### PR TITLE
Feature/token clone factory

### LINKED ISSUE #141: feature: clone
# Motivation
Make deployments cheaper. 
See https://github.com/corpus-io/tokenize.it-smart-contracts/issues/129 and #134 

# Scope

Contracts that would benefit from a clone or proxy factory:
- Token
- ContinuousFundraising
- VestingWallet
- FeeSettings (only if #135 is implemented with multiple FeeSettings contracts)

# Implementation
## Token
- ERC2771Context:
  - trustedForwarder is private immutable -> ok
- ERC20Permit
  - EIP712: name hash and domain separator are private immutable -> **BAD**
  - ERC20: name and symbol are private -> ok
- Pausable
  - paused: private -> ok
- AccessControl
  - mapping roles: private -> ok 

The domain separator of each new token is different from the last one, as it is derived from name, address, chain id and version. The current implementation defines it as private immutable though, which means it is calculated at contract creation and then each occurrence in the code [is replaced with the resulting value](https://docs.soliditylang.org/en/v0.6.5/contracts.html#constant-and-immutable-state-variables). Once the contract is deployed, there is no "variable" anymore, instead this value has become an integral part of the bytecode.

This means that each clone or proxy would use the same value, which is read from the logic contracts code. This is not acceptable for the domain separator in EIP712. We would have to rewrite ERC20Permit in order to proxy or clone our token.

EUROC for example did [their own implementation of EIP712](https://vscode.blockscan.com/ethereum/0xa2327a938febf5fec13bacfb16ae10ecbc4cbdcf), which is assigned through the `initializeV2` function.:
 
```solidity
contract EIP712Domain {
    /**
     * @dev EIP712 Domain Separator
     */
    bytes32 public DOMAIN_SEPARATOR;
}
```

### Gas consumption measurements
| transaction      | natively deployed | cloned |
| ----------- | ----------- | --- |
| transfer      |   26642     | 26744 |
| deployment   |    2.7M (2710487)  | 0.3M (330629) |

A new test for this purpose yields higher deltas for some reason, but support the hypothesis that the "per transaction cost increase of using clones" is largely constant:

>   gas used for transfer: 26393 vs 26209, delta - 184
>   gas used for minting allowance: 25376 vs 25201, delta - 175
>   gas used for grant role: 28901 vs 28726, delta - 175

An economical exploration based on these findings was conducted [here](https://docs.google.com/spreadsheets/d/1S-RhTAt8sfNqwdBPmHY8_C5YfBR7EwWF3ERvWAY9sNM/edit?usp=sharing).

## PersonalInvite
As PersonalInvite stores no code on-chain, there is no point in cloning it.
Update: they do store code! Why? 

Example: https://etherscan.io/address/0xFCAa68aE468bE089687Ad042F02C9050836b3325#code

Solution: Reduced the deployment size a lot by [optimizing through the yul pipeline and disposing of the checksum metadata](https://github.com/corpus-io/tokenize.it-smart-contracts/pull/167).

## ContinuousFundraising
Inherits from
- ERC2771: ok, see above
- Ownable2Step: private _pending -> ok
- Pausable: ok, see above
- ReentrancyGuard: private _status -> ok

Should be cloneable.

## VestingWallet
Has 3 private immutable declarations:
- _beneficiary
- _start
- _duration

-> not cloneable without major changes

### PR DESCRIPTION
**Pull request by malteish**
closes #141 
variable declarations in contracts we inherit from prevent cloning. This means we have to change declarations and write some variables using assembly.

Alternatively, I am exploring making the relevant contracts cloneable with initializers - this work is done in a fork of the openzeppelin contracts.

# Relevant parts that are being tested
## Token
- [x] ERC2771
- [x] Pausing
- [x] AccessControl
- [x] Permit -> see #173 

## ContinuousFundraising

- [x] ERC2771
- [x] Ownable
- [x] Pausable
- [x] ReentrancyGuard

### CODE REVIEW

File: contracts/ContinuousFundraising.sol [hunk 1]
```diff
@@ -114,6 +130,19 @@ contract ContinuousFundraising is ERC2771Context, Ownable2Step, Pausable, Reentr
...
+        minAmountPerBuyer = _minAmountPerBuyer;
+        maxAmountPerBuyer = _maxAmountPerBuyer;
+        tokenPrice = _tokenPrice;
+        maxAmountOfTokenToBeSold = _maxAmountOfTokenToBeSold;
+        currency = _currency;
+        token = _token;
+        // Initialize owner by writing the private variable `_owner` of the Ownable contract using assembly.
+        // This will re-write the storage slot if the contract is created through the constructor, which is acceptable.
+        require(_owner != address(0), "owner can not be zero address");
+        assembly {
```
Comment (CJentzsch) [inline contracts/ContinuousFundraising.sol:143]: Why is this necessary?

File: contracts/Token.sol [hunk 2]
```diff
@@ -131,6 +144,54 @@ contract Token is ERC2771Context, ERC20Permit, Pausable, AccessControl {
 
         // set requirements (can be 0 to allow everyone to send and receive tokens)
         requirements = _requirements;
+
+        // force-initialize the private variables name and symbol inherited from ERC20 using inline assembly
+        // solhint-disable-next-line no-inline-assembly
+        _storeString(3, _name);
```
Comment (CJentzsch) [inline contracts/Token.sol:150]: I really don't like that. Is this really necessary?

File: contracts/Token.sol [hunk 3]
```diff
@@ -131,6 +144,54 @@ contract Token is ERC2771Context, ERC20Permit, Pausable, AccessControl {
...
+        _storeString(4, _symbol);
+    }
+
+    /**
+     * Taken from https://ethereum.stackexchange.com/questions/126269/how-to-store-and-retrieve-string-which-is-more-than-32-bytesor-could-be-less-th
+     */
+    function _storeString(uint256 _lengthSlot, string memory _newContent) private {
```
Comment (CJentzsch) [inline contracts/Token.sol:157]: If we could avoid that, that would be great.

**Review by CJentzsch — CHANGES_REQUESTED** [review]

File: contracts/ContinuousFundraising.sol (hunk 1 above)
Comment (malteish) [reply contracts/ContinuousFundraising.sol:143]: If the contract is created through the constructor, openzeppelin's ownable does this for us:
`require(newOwner != address(0), "Ownable: new owner is the zero address");`
We should provide the same protection when creating a clone.

File: contracts/Token.sol (hunk 3 above)
Comment (malteish) [reply contracts/Token.sol:157]: see above
